### PR TITLE
Add experimental WebGPU HAL backend.

### PR DESCRIPTION
**Pull request by ScottTodd**
This merges the `webgpu` development branch after over a year of on and off development.

For now, this code will live in the `experimental/` folder, while the code matures and reaches feature parity with the supported HAL drivers. While some work has been started to add unit and integration tests for the web platform, that work is not complete yet (and it will likely be substantially different from our existing "native" tests). See also the discussion on [RFC: Promoting IREE web platform / Emscripten builds to "stable"](https://groups.google.com/g/iree-discuss/c/2K5VJ9P8K8I/m/elidhfFMCAAJ).

Major components:

* A runtime HAL driver in `experimental/webgpu/` that targets the recently released [WebGPU API](https://www.w3.org/TR/webgpu/)
  * This uses [webgpu-headers](https://github.com/webgpu-native/webgpu-headers) to connect with API implementations, either via [Emscripten's library_webgpu.js](https://github.com/emscripten-core/emscripten/blob/main/src/library_webgpu.js) or via other implementations like [Dawn](https://dawn.googlesource.com/dawn/) / [wgpu-native](https://github.com/gfx-rs/wgpu-native) (we originally had a platform switch for these, but dropped that for now)
* A sample application in `experimental/web/sample_webgpu/`
  * Build script compiles programs with `--iree-hal-target-backends=webgpu`
  * Build script configures CMake with `-DIREE_EXTERNAL_HAL_DRIVERS=webgpu`
  * Code shows how to interface between runtime C code, browser JavaScript APIs, and Emscripten

### CODE REVIEW

**Review by benvanik — APPROVED** [review]
Gonna say we just land it and then do some cleanup passes once in-tree.

**Review by benvanik — APPROVED** [review]
Gonna say we just land it and then do some cleanup passes once in-tree.

**Comment by ScottTodd** [comment]
Tested that `./experimental/web/testing/build_tests.sh` still works with this (and pushed one fix :P). Going to wait until tomorrow to merge.

**Pull request merged**
